File: daemon/receive/dispatch.py
"""Type -> save + open by association (CLAUDE.md section 3, receive path).

    url   -> open in the default browser
    image -> save to disk -> preview pop (no auto-open; you asked for a look)
    file  -> save to disk -> open with the OS default app
    text  -> write to the clipboard -> pop

The whole point: the sender already decided the type, so this side never asks
"what app opens a .docx". It hands the path to the OS and lets the file
association answer. That's why one dispatcher covers a PDF, a spreadsheet and a
photo without knowing anything about any of them.
"""
import base64
import os
import subprocess
import sys
import webbrowser

import pyperclip

import config
from . import toast


def open_path(path):
    """Open a saved file with whatever the OS says owns that extension."""
    if hasattr(os, "startfile"):                     # Windows
        os.startfile(str(path))
    else:                                            # macOS / Linux
        opener = "open" if sys.platform == "darwin" else "xdg-open"
        subprocess.Popen([opener, str(path)])


def dispatch(env, save_dir, log=print):
    """Act on one received payload. Returns the saved Path, or None."""
    kind = env.get("type")
    sender = env.get("sender", "?")
    data = env.get("data")
    if data is None:
        log(f"dispatch: payload from {sender} has no data — ignored")
        return None

    if kind == "text":
        return _text(data, sender, log)
    if kind == "url":
        return _url(data, sender, log)
    if kind in ("image", "file"):
        return _blob(env, kind, data, sender, save_dir, log)
    log(f"dispatch: unknown payload type {kind!r} from {sender} — ignored")
    return None


def _text(data, sender, log):
    try:
        pyperclip.copy(data)
    except Exception as e:
        log(f"clipboard write failed: {e}")
    toast.show(f"caught: {data[:40]}")
    log(f"CAUGHT text <- {sender}: {data[:60]!r}")
    return None


def _url(data, sender, log):
    # Only ever hand http(s) to the OS. A browser-internal link (chrome://...)
    # pops "Get an app to open this link", and an exotic scheme would launch
    # whatever protocol handler happens to be registered for it — not something
    # to do with a string that arrived over the network. Keep it as text.
    if not data.lower().startswith(("http://", "https://")):
        log(f"CAUGHT url <- {sender}: {data[:60]!r} isn't openable, copied instead")
        return _text(data, sender, log)
    webbrowser.open(data)
    toast.show(f"caught: {data[:40]}", on_click=lambda: webbrowser.open(data))
    log(f"CAUGHT url <- {sender}: {data}")
    return None


def _blob(env, kind, data, sender, save_dir, log):
    try:
        raw = base64.b64decode(data)
    except Exception as e:
        log(f"dispatch: bad base64 in {kind} from {sender}: {e}")
        return None

    name = config.safe_filename(env.get("filename"),
                                default=f"yoink-{kind}")
    path = config.unique_path(save_dir, name)
    path.write_bytes(raw)
    log(f"CAUGHT {kind} <- {sender}: saved {path} ({len(raw)} bytes)")

    if kind == "image":
        toast.show(f"caught: {path.name}", image_path=path,
                   on_click=lambda: open_path(path))
    else:
        toast.show(f"caught: {path.name}", on_click=lambda: open_path(path))
        open_path(path)                              # let the association decide
    return path

File: mobile/lib/receive/dispatch.dart
/// Type -> save + open by association (DESIGN.md section 3, receive path).
///
/// Port of daemon/receive/dispatch.py:
///     url   -> open in the default browser
///     image -> save to disk -> preview pop (tap to open full size)
///     file  -> save to disk -> open with the OS default app
///     text  -> write to the clipboard -> pop
///
/// The whole point: the sender already decided the type, so this side never
/// asks "what app opens a .docx". It hands the file to Android and lets the
/// MIME association answer. One dispatcher covers a PDF, a spreadsheet and a
/// photo without knowing anything about any of them.
///
/// Split into two phases, unlike the desktop's single dispatch():
///   [prepare] runs on arrival — decode and save, so the pop can show a real
///             thumbnail — but opens nothing.
///   [open]    runs on the RECEIVE gesture, or when you tap the pop.
/// The desktop can fuse these because a Tk pop is cheap to build from a path;
/// here the pop is a widget that needs the saved file up front. Keeping the
/// open separate also means an arriving payload never launches an app behind
/// your back — it waits for the gesture, exactly like the desktop's camera path.
library;

import 'dart:convert';
import 'dart:io';

import 'package:flutter/services.dart';
import 'package:open_filex/open_filex.dart';
import 'package:url_launcher/url_launcher.dart';

import 'paths.dart';

/// One received payload, decoded and saved, ready to be opened.
class Caught {
  const Caught({
    required this.type,
    required this.label,
    required this.sender,
    this.file,
    this.text,
    this.url,
    this.mime,
  });

  final String type; // text | url | image | file
  final String label; // what the pop shows
  final String sender;
  final File? file; // set for image/file
  final String? text; // set for text (and for a url we refused to launch)
  final String? url; // set for a launchable http(s) url
  final String? mime; // from the envelope; null -> guess from the extension

  bool get isImage => type == 'image' && file != null;
}

/// Decode and save one arriving payload. Returns null if it isn't usable.
Future<Caught?> prepare(
  Map<String, dynamic> env, {
  required Directory saveDir,
  required void Function(String) log,
}) async {
  final kind = env['type'] as String?;
  final sender = (env['sender'] as String?) ?? '?';
  final data = env['data'] as String?;
  if (data == null) {
    log('dispatch: payload from $sender has no data — ignored');
    return null;
  }

  switch (kind) {
    case 'text':
      return Caught(
        type: 'text',
        label: _short(data),
        sender: sender,
        text: data,
      );

    case 'url':
      // Only ever hand http(s) to the OS. An exotic scheme would launch
      // whatever handler happens to be registered for it — not something to do
      // with a string that arrived over the network. Keep it as text.
      final lower = data.toLowerCase();
      if (!lower.startsWith('http://') && !lower.startsWith('https://')) {
        log('CAUGHT url <- $sender: ${_short(data)} isn\'t openable, copied instead');
        return Caught(
          type: 'text',
          label: _short(data),
          sender: sender,
          text: data,
        );
      }
      return Caught(
        type: 'url',
        label: _short(data),
        sender: sender,
        url: data,
      );

    case 'image':
    case 'file':
      final Uint8List raw;
      try {
        raw = base64Decode(data);
      } on FormatException catch (e) {
        log('dispatch: bad base64 in $kind from $sender: $e');
        return null;
      }
      final name = safeFilename(
        env['filename'] as String?,
        fallback: 'yoink-$kind',
      );
      final file = uniquePath(saveDir, name);
      await file.writeAsBytes(raw);
      final shortName = file.path.split('/').last;
      log('CAUGHT $kind <- $sender: saved $shortName (${raw.length} bytes)');
      return Caught(
        type: kind!,
        label: shortName,
        sender: sender,
        file: file,
        mime: env['mime'] as String?,
      );

    default:
      log('dispatch: unknown payload type $kind from $sender — ignored');
      return null;
  }
}

/// Act on a prepared payload: clipboard, browser, or the OS file association.
/// Returns a short line describing what happened, for the log and the pop.
Future<String> open(Caught caught, {required void Function(String) log}) async {
  switch (caught.type) {
    case 'text':
      await Clipboard.setData(ClipboardData(text: caught.text!));
      log('CAUGHT text <- ${caught.sender}: copied to clipboard');
      return 'copied to clipboard';

    case 'url':
      final uri = Uri.parse(caught.url!);
      final ok = await launchUrl(uri, mode: LaunchMode.externalApplication);
      log('CAUGHT url <- ${caught.sender}: ${ok ? 'opened' : 'no browser found'}');
      return ok ? 'opened in browser' : 'no app could open that link';

    case 'image':
    case 'file':
      // The envelope's mime wins; when it's absent, passing null lets
      // open_filex map the extension to a MIME type itself, so a payload that
      // arrived without one still opens.
      final res = await OpenFilex.open(caught.file!.path, type: caught.mime);
      log('CAUGHT ${caught.type} <- ${caught.sender}: open -> ${res.message}');
      return switch (res.type) {
        ResultType.done => 'opened ${caught.label}',
        ResultType.noAppToOpen => 'saved — no app installed for this type',
        ResultType.fileNotFound => 'saved file went missing',
        ResultType.permissionDenied => 'Android refused to open it',
        ResultType.error => 'could not open: ${res.message}',
      };

    default:
      return 'nothing to open';
  }
}

String _short(String s) => s.length > 40 ? '${s.substring(0, 40)}...' : s;
